### PR TITLE
[WIP] Retinanet Training

### PR DESCRIPTION
**Pull request by mszsorondo**
next (high level) steps: 

- [x] L1 loss
- [x] focal loss for anchor classifications
- [x] L1 loss for box regressions
- [x] clean up code
- [ ] adapt for mlperf-retinanet training specs
- [ ] check equal training updates
- [ ] refactor for exploiting tensor lazyness and tensor mul + JIT
- [ ] Final refactor

### CODE REVIEW

File: extra/training.py [hunk 1]
```diff
@@ -13,6 +13,24 @@ def sparse_categorical_crossentropy(out, Y):
...
   return out.mul(y).mean()
 
+def focal_loss(p : Tensor, y_train : Tensor, 
+               alpha: Tensor = Tensor(0.25, dtype=dtypes.float32), 
+               gamma: Tensor = Tensor(2, dtype=dtypes.float32), 
+               reduction: str = "mean") -> Tensor:
+    p_t = p * y_train + ((Tensor.ones_like(p) - p) * (Tensor.ones_like(y_train) - y_train))
+    ce_loss = -(p_t + 1e-10).log()
+    loss = ce_loss.mul((Tensor.ones_like(p_t) - p_t) ** (gamma))
+    if alpha >= 0:
```
Comment (ZenginU) [inline extra/training.py:23]: if alpha is tensor this will always be true statement

Comment (mszsorondo) [reply extra/training.py:24]: Fixed! thanks

**Comment by mszsorondo** [comment]
forgot to mention  #1636

**Comment by geohot** [comment]
Not updated within a week, feel free to reopen if you are still working on this

**Comment by mszsorondo** [comment]
> Not updated within a week, feel free to reopen if you are still working on this

yeap. it's been a tough couple weeks, will reopen. made some more advances not yet commitable. thanks

**Comment by wozeparrot** [comment]
Can't really give reopen rights, just open another pr, should be fine.
